### ISSUE TITLE
"Find in Files" fails on Ubuntu 14.04.1

### ISSUE THREAD

**Issue by soloproyectos** [geany/geany#451]
The utility "Find in Files" (Shift+Ctrl+F) fails on Ubuntu 14.04.1 when I try to search a string in the files of a directory. It throws the following error:
```
grep -nHIF -- background (in directory: </home/user/path/to/folder>)
/bin/grep: <folder>: Is a directory
...
Search failed.
```

Tested on Geany 1.23.1 and Ubuntu
Ubuntu 14.04.1 LTS

**Comment by kugel-** [comment]
I can reproduce this. Note this only occurs if recursive find is disabled (I guess not a lot of people do this)

Funnily, if the project file type filter is selected then it works, presumably because it builds a file list before grepping, while the custom file type filter passes --include=XX for each filter token.

**Comment by elextr** [comment]
@kugel- makes sense, greps arguments must be files to search in unless the -r option is given.

GNU grep has --directories= option to say what to do with dirs in the filelist, but thats not POSIX AFAIK.

**Comment by kugel-** [comment]
I guess the solution is to always build a file list if -r is not checked, and build --include=X options if -r is checked.

**Comment by kugel-** [comment]
Actually, --include seems non-posix too, but we already depend on it

**Comment by elextr** [comment]
ummm, yes, good point, but if its what we already do then lets keep doing it :)

**Comment by kugel-** [comment]
Actually I can't reproduce this anymore; both a Geany from august 14 and current git HEAD seems to work (builds a file list for custom and project filters)

**Comment by kugel-** [comment]
Wait wait wait, now I've got this. The initial report is slightly misleading. Actually the error message doesn't appear for the search directory, but for sub-directories in it.

This is because search.c already appends a complete file listing of the search directory to the grep command line. This includes directory entries.

So in essence this issue is not so bad, because the files are still grepped. grep just prints an error for directory entries but continues anyway. So this is merely a warning and find in files does NOT fail.

So we could use the --directories=skip option to grep, or filter out directory entries ourselves with g_file_test. Since we depend on GNU extensions already and BSD grep also seems to support this I'd prefer the first, because grep will do the same check again regardless.

But I do notice there's something in search.c that doesn't work. When --include= is given (either via filters or via extra options) we intent to pre-filter the file list (see https://github.com/geany/geany/blob/master/src/search.c#L1779). But this doesn't seem to work at all, I always get the full file list as arg to grep.

**Comment by HybridDog** [comment]
l get an error message on Ubuntu 16.10, geany 1.28.
Anyway, there's another problem: It searches everything, even if it's not a text file.
So e.g. when there's a fifo in the folder ($ mkfifo somefifo), the search becomes locked and doesn't finish.

**Comment by elextr** [comment]
@HybridDog nothing is locked, grep is patiently waiting for end of file on the fifo and Geany is patiently waiting for grep.

If its decided that only GNU grep is to be used then we could add `--devices=skip` to the grep command.

**Comment by MaximUltimatum** [comment]
I can consistently reproduce this bug on Ubuntu 16.04.6 LTS. The search still completes, but I don't get the number of matches on the bottom that I get when the search succeeds. Is there a way print this even when the search fails?

**Comment by codebrainz** [comment]
> then we could add --devices=skip to the grep command

@HybridDog if that works, you should be able to add the option(s) mentioned above into the "Extra Options" on the Find in Files dialog. A grep option like `-I` might also help with non-text files and/or maybe `--exclude` if you know the specific FIFO filename.
